feat: Add new toolset for OpenSearch Agentic Memory API

oujezdsky opened the pull request:
### Description
This pull request introduces a set of 6 new tools to support the **OpenSearch Agentic Memory API** (released in 3.3). This allows agents using the MCP server to create, manage, and interact with agentic memories.

**Key changes include:**

* **New Tools:** Adds `CreateAgenticMemorySessionTool`, `AddAgenticMemoriesTool`, `GetAgenticMemoryTool`, `UpdateAgenticMemoryArgsTool`, `DeleteAgenticMemoryByIDTool`, `DeleteAgenticMemoryByQueryTool`, and `SearchAgenticMemoryTool`.
* **Pydantic Models:** Implements Pydantic V2 models for all new tool arguments.
* **Async Implementation:** The core logic is implemented as `async` helper functions in `src/opensearch/helper.py`.
* **Error Handling:** Introduces a new `HelperOperationError` to provide better, contextualized error logging.
* **Testing:** Adds a set of unit tests for the new tools, using `pytest.mark.parametrize` for all happy paths and `pytest.raises` for Pydantic validation.


### Issues Resolved
Closes #113

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

oujezdsky committed:
Add tool params for memory tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Added handlers and tools for memory tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Escape user-provided path params

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Improve JSON type annotations

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Fixed HTTP method for memory search, POST -> GET

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Added custom Exception handler, utilized in new memory tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Helper functions reworked

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
tools functions and tool params reworked

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
tests added

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
fixed issues so it works with updated baseToolArgs with opensearch_cluster_name param

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
correctly await async calls in agentic memory helpers

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
apply ruff formatting to agentic memory feature files

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
apply ruff check to agentic memory feature files

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
docs: Update CHANGELOG for Agentic Memory tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
fix(packaging): Move exceptions.py to tools package to fix installation

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Add compatibility checks to Agentic Memory tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
fix(tools): Rename UpdateAgenticMemoryArgsTool to UpdateAgenticMemoryTool

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
fix(tools) Correct agentic memory strategy type field naming

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
fix(CHANGELOG): Rename UpdateAgenticMemoryArgsTool to UpdateAgenticMemoryTool

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky commented:
Hello, 

This PR is ready for review again.

 **5 new commits** that address issues:

* **Fixed packaging** by moving `exceptions.py`.
* **Added compatibility checks** to all Agentic Memory tools.
* **Corrected naming** for `UpdateAgenticMemoryTool` (in code and `CHANGELOG`).
* **Fixed CreateAgenticMemoryContainerTool validation** by changing the strategy field from invalid `strategy_type` to the correct **`type`**.

Thank you for the review and apologies for the inconvenience.

rithin-pullela-aws commented:
Hi @oujezdsky thanks for the PR!

Few high level thoughts before I review the PR:
- Can we fix the failing tests?
- Can we add the tools in Readme and Userguide?
- Looks like the tool_params.py and tool.py have a lot of lines of code added. Can we create separate files for agentic_memory related tools and import the tools into tools.py ? I believe this would make the code more readable

oujezdsky committed:
fix(tools): Rename UpdateAgenticMemoryArgsTool to UpdateAgenticMemoryTool

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
refactor(tools): extract Agentic Memory tools to dedicated package

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
test(agentic-memory): isolate tests and fix OpenSearch 3.3 mocking

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
docs(agentic-memory): document tools, parameters, and usage examples

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

nathaliellenaa commented on src/opensearch/helper.py:
```diff
@@ -292,3 +296,266 @@ async def get_opensearch_version(args: baseToolArgs) -> Version:
+    """Create a new agentic memory session in the specified memory container.
+
+    Args:
+        args: CreateSessionArgs containing memory_container_id and optional session_id, summary, metadata, namespace
```
Nit: 
```suggestion
        args: CreateAgenticMemorySessionArgs containing memory_container_id and optional session_id, summary, metadata, namespace
```

nathaliellenaa commented on src/tools/agentic_memory/params.py:
```diff
@@ -0,0 +1,873 @@
+    @model_validator(mode='after')
+    def validate_embedding_configuration(self) -> 'AgenticMemoryConfigurationArgs':
+        """Validate embedding model configuration."""
+        if self.embedding_model_type == 'TEXT_EMBEDDING' and self.embedding_dimension is None:
```
Can we do `if self.embedding_model_type == EmbeddingModelType.text_embedding`?

nathaliellenaa reviewed (commented):
Thanks for raising this PR @oujezdsky! Added some comments

oujezdsky committed:
docs(agentic-memory): apply review feedback to README and CHANGELOG

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
fix(agentic-memory): correct http_methods in TOOL_REGISTRY

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
style(agentic-memory): fix naming inconsistency

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
refactor(params, tests): prefer Enums over string literals for consistency

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
refactor(actions): consistent message formatting

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
style(agentic-memory): reorder components for consistency

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

jiapingzeng commented:
Thanks for the PR! Overall the changes look good to me from a tools perspective. 

One question is that when using these MCP tools with an agent, e.g. Claude desktop or Q/Kiro CLI, is it able to recognize that it needs to create a container before it can use the other Agentic Memory tools? Or would the user need to guide it through prompt to create container then add/get/update/delete memory?

oujezdsky commented:
Thanks for the feedback!

Regarding the MCP tools themselves, it relies on the tool definitions (schemas). For example, since `AddAgenticMemoriesTool` requires a `memory_container_id`, the user should deduce that the `CreateAgenticMemoryContainerTool` must be called first in order to obtain the `memory_container_id`.

But I could extend the description in `BaseAgenticMemoryContainerArgs`, where the mandatory field `memory_container_id` is specified.

For example, like this:
```
class BaseAgenticMemoryContainerArgs(baseToolArgs):
    """Base arguments for tools operating on an existing Agentic Memory Container."""

    memory_container_id: str = Field(
        ..., 
        description='The ID of the memory container. You must obtain this ID by calling CreateAgenticMemoryContainerTool first.'
    )
```
(snippet from `src/tools/agentic_memory/params.py`)

oujezdsky committed:
Merge remote-tracking branch 'upstream/main' into feature/113-add-memory-tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

dhrubo-os commented:
> Thanks for the feedback!
> 
> Regarding the MCP tools themselves, it relies on the tool definitions (schemas). For example, since `AddAgenticMemoriesTool` requires a `memory_container_id`, the user should deduce that the `CreateAgenticMemoryContainerTool` must be called first in order to obtain the `memory_container_id`.
> 
> But I could extend the description in `BaseAgenticMemoryContainerArgs`, where the mandatory field `memory_container_id` is specified. (`BaseAgenticMemoryContainerArgs` is used only within add, create session, get, update, delete, and search...)"
> 
> For example, like this:
> 
> ```
> class BaseAgenticMemoryContainerArgs(baseToolArgs):
>     """Base arguments for tools operating on an existing Agentic Memory Container."""
> 
>     memory_container_id: str = Field(
>         ..., 
>         description='The ID of the memory container. You must obtain this ID by calling CreateAgenticMemoryContainerTool first.'
>     )
> ```
> 
> (snippet from `src/tools/agentic_memory/params.py`)

I think we should keep container creation separate from the MCP tools for these reasons:

1. Container creation is infrastructure setup - It requires careful configuration of embedding models, LLM connectors, strategies, and index settings. This is typically a one-time admin operation, not something 
agents should do at runtime.

2. Runtime operations need pre-configured containers - For the actual memory operations (add, search, update, delete), agents should use a pre-existing memory_container_id.

3. MCP server already supports this pattern - Users can pre-configure the memory_container_id in the tool schema:
  ```
   tools:
     AddAgenticMemoriesTool:
       args:
         memory_container_id: "your-container-id-here"
 ```
   
  This way, agents don't need to pass it with every call.

nathaliellenaa commented:
> >I think we should keep container creation separate from the MCP tools for these reasons
> 
> +1 on we should keep memory container creation separate as otherwise agent will create a new container each time and won't be able to access previous session's memory.

> Though for ease-of-use purpose we should consider if we can provide an easy way to create the container either via a script or via optional MCP tools. Having to manually setup embedding models, LLM connectors, strategies, index settings manually could be overwhelming for new users of Agentic Memory.


I agree on separating the memory container creation from the MCP tools as the configuration can be complex and it's error prone for agents to handle at runtime.

To make the initial setup easier, we could support pre-built container templates (e.g., default, minimal, advanced). This way, new users who aren't familiar with Agentic Memory can simply run  `python -m mcp_server_opensearch --container_template default`. This would create a container with sensible defaults, and more advanced users can still customize their own setup

> > Users can pre-configure the memory_container_id in the tool schema
> 
> Yes container ID can be provided via tool customization, maybe we can also allow it to be configured via env vars.
> 
> I think memory tools should also be disabled by default (so that agents don't try to invoke them without a container ready), and users who do want to use memory can specify flag to enable and provide memory container ID.

+1 on this. I think we can introduce a dedicated flag for the agentic memory feature. When users enable this flag, all agentic memory-related tools will be automatically enabled, and they can provide the memory_container_id at server startup. For example `python -m mcp_server_opensearch --agentic_memory --memory_container_id <id>`

dhrubo-os commented:
@oujezdsky if possible can we target to make the suggested changes and merge this PR asap? Thanks a lot for working on this.

oujezdsky commented:
Hi everyone, thanks for the detailed feedback.

@dhrubo-os  I'll be able to work on it this weekend. There's a chance I'll be able to do it sooner, but the weekend is the most certain.
I'll push the updates as soon as I have them ready. Thanks

oujezdsky committed:
Merge remote-tracking branch 'upstream/main' into feature/113-add-memory-tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky commented:
Okay

oujezdsky committed:
refactor: remove Create Container Tool

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
feat(config): add support for memory_container_id in config and env

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
docs: add support for memory_container_id in config and env

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
feat: support for memory_container_id, removed faulty runtime default handling; now applied during validation.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
feat: support for memory_container_id, simplified config

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Merge remote-tracking branch 'upstream/main' into feature/113-add-memory-tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

rithinpullela commented:
Hi @oujezdsky can you please resolve the conflicts on this PR?

oujezdsky committed:
Merge remote-tracking branch 'upstream/main' into feature/113-add-memory-tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky commented:
Hello @rithinpullela 
merge is done.

rithinpullela commented on src/tools/agentic_memory/actions.py:
```diff
@@ -0,0 +1,246 @@
+import json
```
Can we add the copyright header? refer to other files

rithinpullela replied:
Same pattern found in other instances with new files as well, can we fix all those cases too?

rithinpullela commented on src/tools/tools.py:
```diff
@@ -894,9 +896,433 @@ async def search_experiments_tool(args: SearchExperimentsArgs) -> list[dict]:
         return log_tool_error('SearchExperimentsTool', e, 'searching experiments')
 
 
+async def create_search_configuration_tool(args: CreateSearchConfigurationArgs) -> list[dict]:
```
Looks like a bad merge? These tools exist from before

rithinpullela replied:
+1 till line 1254, can we fix this?

rithinpullela commented on CHANGELOG.md:
```diff
@@ -28,6 +28,29 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
+
+### Removed
+
+## [Released 0.9.0]
```
A log of duplicated changes can we fix these?

rithinpullela commented on src/tools/agentic_memory/actions.py:
```diff
@@ -0,0 +1,246 @@
+            else f'Session created, but no ID was returned. Response: {json.dumps(result)}'
+        )
+        return [{'type': 'text', 'text': message}]
+    except Exception as e:
```
In case of error can we use the existing patterns in other tools? like `log_tool_error('SearchExperimentsTool', e, 'searching experiments')`

This allows us to make sure we cna have deterministic tool error logging and monitoring if these logs are published into cloudwatch or an equivalent

rithinpullela commented on src/tools/utils.py:
```diff
@@ -3,7 +3,13 @@
 
 import logging
 import yaml
+import logging
```
Duplicated import

rithinpullela commented on src/mcp_server_opensearch/stdio_server.py:
```diff
@@ -39,6 +39,9 @@ async def serve(
     await generate_tools_from_openapi()
+    # Register agentic memory tools if memory_container_id is configured
+    from tools.tools import register_agentic_memory_tools
+    register_agentic_memory_tools(config_file_path)
```
Can we somehow not expose this in the main server file? I feel like this should belong in the internal implementation of get_tool function? Want to hear your thought about this

rithinpullela commented on src/mcp_server_opensearch/tool_executor.py:
```diff
@@ -65,7 +65,7 @@ async def execute_tool(
         from tools.tool_params import validate_args_for_mode
 
-        parsed = validate_args_for_mode(arguments, tool['args_model'])
+        parsed = validate_args_for_mode(arguments, tool['args_model'], tool['input_schema'])
```
I prefer not changing this function, as this can potentially interfere with other tools, can we handle this in memory tool specific methods?

rithinpullela commented on src/tools/tool_params.py:
```diff
@@ -2,29 +2,46 @@
+    args_dict: Dict[str, Any],
+    args_model_class: Type[T],
+    input_schema: Optional[Dict[str, Any]] = None,
+) -> T:
```
This function is meant to validate args based on mode, now the responsibility of this function seems to be increasing, is there a better way to handle the default value for memory ID?

rithinpullela commented on src/opensearch/helper.py:
```diff
@@ -702,6 +706,244 @@ async def get_opensearch_version(args: baseToolArgs) -> Version:
+        try:
+            response = await client.transport.perform_request(method='POST', url=url, body=body)
+        except Exception as e:
+            raise helper_error('create agentic memory session', e)
```
Can we follow error handling from other tools instead of creating infra changes(like defining new exception types)? 

This helps us keep the PR focused

rithinpullela reviewed (commented):
Hi @oujezdsky thanks for the PR!

I want to start off by saying this is great work! The feature is solid and tools work well. I really like the behavior where when we provide the memory container ID in env variables it is automatically made into not required tool, great addition there! I also like the Pydantic @model_validator cross-field validation!

I do have some comments related to changes to the other parts of the code base. I would like to keep the addition of new tools focused on adding the new tools without changing much of the other parts of the code base and it would be great if we can follow the patterns established in the code base. 

I believe there is insufficient guidance in the repo about adding new tools and hence it is very hard to follow to patterns of this code base. Acknowledging that I raised a PR to add more guidance on how we can add new tools to this repo following some standard patterns. You can refer to them [here](https://github.com/rithinpullela/opensearch-mcp-server-py/blob/007986016fcee6848dea260cd911452b6086dcd7/DEVELOPER_GUIDE.md#adding-custom-tools). I raised a PR to update the [dev guide](https://github.com/opensearch-project/opensearch-mcp-server-py/pull/214) and add an agents.md file

**Quick overview of pattern to follow:**
Every tool follows a 4-piece pattern: Pydantic params (extends `baseToolArgs`), async helper (no try/except lets exceptions propagate), async tool function (catches with `log_tool_error()`), static `TOOL_REGISTRY` entry. The best reference is the SRW tools 18 tools added across PRs #187, #190, #192, #193 with zero infrastructure changes.

### Review/ Feedback:

We have this thing called "tool categories" through which we can group similar tools into categories and users can enable these tools by configuring through yaml or env variables. The SRW(search relevance workbench) tools follow this pattern where they define 'search_relevance' category and these tools will be enabled from user config. Can we adopt this for Agentic Memory tools as well? TO make it less confusing to the end user we should not enable memory tools based on the presence of `memory_container_id` variable.


**Use log_tool_error() for error handling** : Every existing tool uses it from `tool_logging.py`. It returns `is_error: True`, which `tool_executor.py `(line 75) checks for structured metrics. Without it, errors are silently counted as successes. This also eliminates the need for `exceptions.py`, `HelperOperationError`, and `helper_error()`.

**Helpers shouldn't catch exceptions**: All existing helpers are "dumb pipes"  make the call, let exceptions propagate. The tool function handles errors.

**Add agentic_memory category to tool_filter.py** : Currently the memory tools aren't in any category (core_tools or search_relevance). The filter enables core_tools by default and filters out everything else. Without a category, these tools are invisible even when registered. Follow the search_relevance pattern.

**Don't modify validate_args_for_mode()** : This is a shared hot path for every tool call. Can you see if we can replace this with the method I used in my draft PR below?

**Registration**: No need for `register_agentic_memory_tools()` in the startup pipeline. Put tools in TOOL_REGISTRY (or spread a separate registry dict like skills_tools.py does) and control visibility through the category system.

I put together a draft PR that restructures this to follow the established patterns, please refer to it, it can be helpful. [Draft PR](https://github.com/opensearch-project/opensearch-mcp-server-py/pull/215) for your reference.

oujezdsky commented:
Hello @rithinpullela, thanks for the feedback—I'll take a look at it on Sunday

dhrubo-os commented:
> Hello @rithinpullela, thanks for the feedback—I'll take a look at it on Sunday

Is there any update @oujezdsky ? We need this feature asap. will this be possible to wrap this feature by this week? Please lmk if we can unblock you somehow. Thanks.

oujezdsky commented:
@dhrubo-os 
yes - this week is real
I've already got a larger part done, but I've been really busy with work these past few weeks, so it's going slowly.

oujezdsky committed:
refactor: remove custom exception infrastructure

Remove HelperOperationError, helper_error() utility, and the
exceptions module. Error handling will use the existing
log_tool_error() pattern from tool_logging.py instead.

Addresses review feedback to follow established error handling
conventions without introducing new exception types.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
refactor(agentic-memory): adopt log_tool_error and define TOOL_REGISTRY

Rewrite tool functions to use log_tool_error() from tool_logging.py
instead of custom HelperOperationError handling. This preserves
is_error: True for structured metrics in tool_executor.py.

Move AGENTIC_MEMORY_TOOLS_REGISTRY into the actions module following
the skills_tools.py spread pattern. Add copyright header.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
refactor(agentic-memory): make helpers propagate exceptions

Remove try/except blocks from all agentic memory helper functions
so they act as dumb pipes. Exceptions now propagate to the tool
functions which handle them with log_tool_error().

Replace wildcard imports with explicit named imports.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
refactor(agentic-memory): inject memory_container_id via model_validator

Use Pydantic @model_validator on BaseAgenticMemoryContainerArgs to
inject memory_container_id from config/env when not provided by the
MCP client. This replaces the previous approach of modifying
validate_args_for_mode() and keeps the injection self-contained
within the agentic memory module.

Remove should_enable_agentic_memory_tools() as tool visibility is now
controlled through the category system. Add copyright headers.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
refactor: use TOOL_REGISTRY spread, remove dynamic registration

Replace wildcard imports and register_agentic_memory_tools() with
the **AGENTIC_MEMORY_TOOLS_REGISTRY spread pattern, matching
how SKILLS_TOOLS_REGISTRY is integrated. This removes the need
for startup-time registration calls in server files.

Revert validate_args_for_mode() to its original signature by
removing the input_schema parameter, as memory_container_id
defaults are now handled via Pydantic model_validator.

Remove duplicate tool function definitions from bad merge.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
feat(tool-filter): add agentic_memory category

Register agentic memory tools as a built-in category following
the search_relevance pattern. The category is not enabled by
default and must be explicitly enabled via enabled_categories
in YAML config or environment variables.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
docs: update CHANGELOG for agentic memory tools

Add agentic memory toolset entry to Unreleased section. Fix
duplicate entries from bad merge in Released sections.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
test(agentic-memory): update tests for log_tool_error error handling

Update test assertions to match the new log_tool_error() response
format which sets is_error: True on error responses. Adjust
expected message strings to match simplified tool output.

Remove register_agentic_memory_tools() call from test setup as
tools are now statically registered via TOOL_REGISTRY spread.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
Merge remote-tracking branch 'upstream/main' into feature/113-add-memory-tools

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

# Conflicts:
#	CHANGELOG.md
#	example_config.yml
#	src/tools/tool_filter.py

oujezdsky commented:
Hi @rithinpullela , thanks for the detailed review and the reference PR! I've addressed all the feedback:

- Replaced custom exceptions with log_tool_error() from tool_logging.py
- More straightforward helpers (no try/except)
- Added agentic_memory category to tool_filter.py (follows search_relevance pattern)
- Reverted changes to validate_args_for_mode() — container ID injection now uses a Pydantic @model_validator on BaseAgenticMemoryContainerArgs
- Switched to **AGENTIC_MEMORY_TOOLS_REGISTRY** spread in TOOL_REGISTRY, removed register_agentic_memory_tools()
- Merged latest upstream/main

All 7 tools tested end-to-end against OpenSearch 3.3.0

rithinpullela commented:
Hi @oujezdsky there seem to be some merge conflicts can you resolve them?

oujezdsky commented:
> Hi @oujezdsky there seem to be some merge conflicts can you resolve them?

Oh, sorry, I forgot to push the changes last time....

rithinpullela commented on CHANGELOG.md:
```diff
@@ -73,6 +74,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Fix search index tool time format issue. ([#159](https://github.com/opensearch-project/opensearch-mcp-server-py/pull/159))
 
 ### Removed
+- Remove `CreateAgenticMemoryContainerTool` - container creation is an infrastructure setup operation that should be done via the OpenSearch API or dashboard, not at agent runtime ([#138](https://github.com/opensearch-project/opensearch-mcp-server-py/pull/138))
```
Looks like this was added by mistake?

rithinpullela commented on README.md:
```diff
@@ -63,6 +63,18 @@ The following tools are available but disabled by default. To enable them, see t
 - [GetLongRunningTasksTool](https://docs.opensearch.org/latest/api-reference/cat/cat-tasks/): Gets information about long-running tasks in the cluster, sorted by running time in descending order.
 
+### Agentic Memory Tools (Disabled by Default)
+The following tools provide AI agents with persistent memory capabilities using the [OpenSearch Agentic Memory API](https://docs.opensearch.org/latest/ml-commons-plugin/agentic-memory/). These tools require OpenSearch **3.3.0 or later** and are **disabled by default**. They are automatically enabled when a `memory_container_id` is configured via the `agentic_memory` section in the config file or  ...
```
I believe we changed this right? We need to enable the `agentic_memory` category to enable these tools. 
Can we also document all of these tools are under `agentic_memory` category?

rithinpullela commented on src/tools/utils.py:
```diff
@@ -6,6 +6,9 @@
 
+# Configure logging
+logger = logging.getLogger(__name__)
+
```
nit: Looks like we are not suing this anymore?

rithinpullela commented on src/tools/agentic_memory/params.py:
```diff
@@ -0,0 +1,653 @@
+    semantic = 'SEMANTIC'
+    user_preference = 'USER_PREFERENCE'
+    summary = 'SUMMARY'
+
```
Are we using these?

rithinpullela commented on src/tools/agentic_memory/params.py:
```diff
@@ -0,0 +1,653 @@
+        default=None,
+        description='Updated structured data content (for data memory payloads).',
+    )
+    binary_data: Optional[Dict[str, Any]] = Field(
```
nit: Binary data is defined as a dict here and as a string in Add Memories tool, should we change this to string?

rithinpullela reviewed (commented):
Thanks for addressing all of my comments @oujezdsky!

I do not have any major concerns. Added few small comments

Also, can you create an issue to add ITs for Agentic Memory? Don't want to block this PR

oujezdsky committed:
docs: fix category enablement documentation

Update README, USER_GUIDE, and CHANGELOG to accurately describe that
agentic memory tools are enabled via the agentic_memory category
(enabled_categories), not by the presence of memory_container_id.

Remove misplaced CHANGELOG entry from Released 0.7.0 section. Fix
missing blank lines in markdown.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

oujezdsky committed:
fix(agentic-memory): address review nits

Remove unused logger from utils.py. Remove EmbeddingModelType and
StrategyType enums from params.py (leftover from removed
CreateAgenticMemoryContainerTool). Fix binary_data type in
UpdateAgenticMemoryArgs to str for consistency with AddAgenticMemoriesArgs.

Signed-off-by: Jiri Oujezdsky <oujezdsky2@gmail.com>

rithinpullela reviewed (approved):
Thanks for addressing the comments! Great work @oujezdsky, this was a complicated feature!

oujezdsky commented:
Thanks @rithinpullela for your help, and to everyone else who participated
